add(explanation): Priority inversion and inheritance (#58)
canonical/real-time-ubuntu-docs: docs/explanation/priority-inversion-inheritance.rst

Priority inversion and inheritance
==================================

> Among the reasons why PREEMPT_RT is not in mainline yet is that much of the locking within the kernel has to be updated to prevent priority inversion from occurring in a real-time environment.

Reviewer: Hi @farshidtz , the intro to PREEMPT_RT here is a bit sudden. I know this is from the blog but at that point in the blog there was more context around the real-time kernel itself and PREEMPT_RT.

I suggest prefacing this a bit with a tiny overview of PREEMPT_RT.

Something along the lines of (not the best, please revise as needed):

The PREEMPT_RT patch is a set of Linux kernel patches with the goal of transforming Linux into a real-time system.

Author: Good point. I've replaced the intro.

The present and the following section will introduce unbounded priority inversion and the need for priority inheritance.

Unbounded priority inversion
----------------------------

Let's begin with priority inversion by looking at the diagram below.
Three tasks, L, M, and H, with varying priority levels, low, medium and high, are present in the kernel and about to contest for CPU access.

The low-priority task L runs until it takes a lock; in the diagram below, the blue bar turns purple.
After acquiring it, task L holds the lock and begins modifying some critical sections within the kernel. 
Once the higher-priority task H appears, it preempts task L and starts to run.

> At this point, task H would like to acquire the same lock task L is holding.

Reviewer: ```suggestion
At this point, task H tries to acquire the same lock task L is holding.
```

As it can't do so, the higher-priority task H goes to sleep and waits for the lower-priority task L to release the lock.

> Task L thus continues running while Task H is sleeping.

Reviewer: ```suggestion
Task L thus continues running while task H is sleeping.
```

In such a setup, priority inversion can occur if a medium-priority task M comes along and preempts task L.
Once task M starts running, the high-priority task H will potentially wait for an unbounded amount of time, preventing it from doing work in a critical kernel section.
Improving the flexibility to preempt tasks executing within the kernel would thus help guarantee an upper time boundary.

.. figure:: priority-inversion-inheritance/unbounded-priority-inversion.svg
    :width: 100%
    :align: center
    :alt: Unbounded priority inversion

>     Unbounded priority inversion: A high-priority task being indefinitely delayed by a medium-priority task due to a lack of proper priority management.

Reviewer: By the way, this isn't part of this PR's scope.
But we might want to make some changes to the CSS for captions. I think the caption text should look a little different from normal paragraph text.
Personal preference but calling this out anyway.

> In this specific example, task M finishes running and releases the CPU -- where the horizontal bar turns from green to red in the drawing -- allowing task L to start running again while still holding the lock.

Reviewer: ```suggestion
In this specific example, task M finishes running and releases the CPU -- where the horizontal bar turns from green to red in the diagram -- allowing task L to start running again while still holding the lock.
```

Nitpick.

Only once task L releases it, task H will wake up and acquire the lock, starting its work within the critical section.

Priority inversion occurred on the Mars Rover, and it is a critical challenge for developers and engineers working with real-time systems. With unbounded priority inversion, the need for priority inheritance becomes clear.

Priority Inheritance
--------------------

A real-time Linux kernel resolves the unbounded latencies of priority inversion via priority inheritance. 

The diagram below helps illustrate the mechanism.
As before, the low-priority task L starts running and acquires the lock.

> Similarly to the previous scenario, task H wakes up and starts running, but it is soon blocked while attempting to get the lock.

Reviewer: ```suggestion
Similar to the previous scenario, task H wakes up and starts running, but it is soon blocked while attempting to get the lock.
```

The high-priority task H wants to take the same lock held by the low-priority task L.

> Differently than in the priority inversion's case, and instead of H going to sleep and waiting, priority inheritance occurs, with L acquiring H's priority.

Reviewer: ```suggestion
In this case, instead of H going to sleep and waiting, priority inheritance occurs, with L acquiring H's priority.
```

This is a suggestion, please feel free to leave as is, but the sentence didn't seem to flow very well IMO.

Thanks.

The low-priority task L can now run with the same priority as task H, enabling it to finish its work in the critical section and then release the lock.
The inheritance mechanism centers around boosting the lower task's priority, giving it one higher than the upcoming medium priority task M, which would cause unbounded latencies.

Once task L finishes its critical section work, task H acquires the lock, where the red bar turns orange.

> Whenever task H completes, it will, in turn, release the lock.

Reviewer: ```suggestion
When task H completes, it will, in turn, release the lock.
```
